Add QgsRectangle::setNull(), deprecating setMinimal()

strk opened the pull request:
Reduces confusion, as noted by @nyalldawson in https://github.com/qgis/QGIS/pull/54917#discussion_r1355849043

strk committed:
Add QgsRectangle::setNull(), deprecating setMinimal()

lnicola commented on src/core/geometry/qgsrectangle.h:
```diff
@@ -165,16 +165,24 @@ class CORE_EXPORT QgsRectangle
+    void setNull() SIP_HOLDGIL
+    {
+      mXmin = mYmin = std::numeric_limits<double>::max();
+      mXmax = mYmax = -std::numeric_limits<double>::max();
```
Nit: I'd slightly prefer `std::numeric_limits<double>::lowest()` if available (C++11).

strk replied:
is there an ::highest() too ?

lnicola replied:
That's called `max` :slightly_smiling_face:.

strk replied:
in any case @benoitdm-oslandia  [suggested to use NaN](https://github.com/qgis/QGIS/pull/54646#discussion_r1330107062) so this is still I think a detail that shouldn't be handled in this PR which is purely for the API part.

lnicola replied:
I guess it depends on how exciting you want calling `center` on a null `QgsRectangle` or doing an `intersect` between a null and a non-null one to be.

strk replied:
I would change semantic of "isEmpty()" to mean "it represents an no space at all" (what isNull() currently means) and drop isNull/setNull all togheter. It would probably be the easiest to explain. I dubt any user is relying on "isEmpty()" meaning "area == 0"

lnicola commented on python/core/auto_generated/geometry/qgsrectangle.sip.in:
```diff
@@ -111,10 +111,18 @@ Set the minimum y value.
-    void setMinimal() /HoldGIL/;
+    void setNull() /HoldGIL/;
+%Docstring
+Mark a rectangle as being null.
```
Might be worth giving more details here about what null means (like `setMinimal` has).

strk replied:
I've force pushed a commit adding more documentation and also the missing \since indication

strk committed:
Rename QgsBox3d::setMinimal() to setNull()

The method was introduced in this version so it should be ok to
rename with no formal deprecation notice

strk committed:
Make TestQgsMeshLayer::test_extent asserts more readable

Expected/Obtained was swapped, see
https://github.com/qgis/QGIS/actions/runs/6518971706/job/17706890258?pr=54934#step:13:352

strk commented:
I went with default-constructing QgsRectangle as null. This PR is now pretty much overlapped with GH-54646 and maybe I should go back to merge them.

nyalldawson commented on src/core/geometry/qgsrectangle.h:
```diff
@@ -165,16 +165,31 @@ class CORE_EXPORT QgsRectangle
+     * \deprecated since QGIS 3.34 - will be removed in QGIS 4.0. Use setNull().
      */
-    void setMinimal() SIP_HOLDGIL
+    Q_DECL_DEPRECATED void setMinimal() SIP_HOLDGIL
```
```suggestion
    Q_DECL_DEPRECATED void setMinimal() SIP_HOLDGIL SIP_DEPRECATED
```

strk replied:
Done, thanks

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -233,6 +233,9 @@ QgsGeometry QgsGeometry::fromMultiPolygonXY( const QgsMultiPolygonXY &multipoly
 
 QgsGeometry QgsGeometry::fromRect( const QgsRectangle &rect )
 {
+  if ( rect.isNull() )
```
Let's leave this for the other pr, and just get the function rename only bit merged first

nyalldawson commented on src/core/geometry/qgsrectangle.h:
```diff
@@ -334,6 +349,8 @@ class CORE_EXPORT QgsRectangle
     */
     QgsRectangle buffered( double width ) const
     {
+      if ( isNull() )
```
As above -- leave this out of the pr

nyalldawson commented on src/core/geometry/qgsrectangle.h:
```diff
@@ -165,16 +165,31 @@ class CORE_EXPORT QgsRectangle
+     */
+    void setNull() SIP_HOLDGIL
+    {
+      mXmin = mYmin = std::numeric_limits<double>::max();
```
Let's leave this out of the renaming pr too

nyalldawson commented on src/core/geometry/qgsrectangle.h:
```diff
@@ -622,10 +649,10 @@ class CORE_EXPORT QgsRectangle
-    double mYmin = 0.0;
-    double mXmax = 0.0;
-    double mYmax = 0.0;
+    double mXmin = std::numeric_limits<double>::max();
```
As above

strk commented:
Ok I've removed all commits not related to the deprecation of setMinimal by setNull now

strk commented:
See how POLYGON EMPTY is handled better than NULL GEOMETRY:

```
QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:451 : (prepare) [2ms] Visible extent: Null
QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:452 : (prepare) [0ms] mapSetting extentBuffer: 0 
QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:454 : (prepare) [0ms] Grown visible extent: Null
QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:456 : (prepare) [0ms] Extent geom  
FAIL!  : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() Compared values are not the same
   Actual   (provider->mLabels.size()): 1
   Expected (0)                       : 0
   Loc: [/usr/local/src/qgis/qgis/src/rectangle-null/tests/src/core/testqgslabelingengine.cpp(955)]
```

vs.
```
90: QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:451 : (prepare) [2ms] Visible extent: Null
90: QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:452 : (prepare) [0ms] mapSetting extentBuffer: 0
90: QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:454 : (prepare) [0ms] Grown visible extent: Null
90: QDEBUG : TestQgsLabelingEngine::testRegisterFeatureUnprojectible() src/core/labeling/qgspallabeling.cpp:456 : (prepare) [0ms] Extent geom  Polygon EMPTY
90: PASS   : TestQgsLabelingEngine::testRegisterFeatureUnprojectible()
```

nyalldawson commented:
Thank you!
